### PR TITLE
N°5090 Setup : improve missing dependencies message

### PR DESCRIPTION
**Pull request by piRGoif**
To reproduce, simply add a single module that is part of a multi module extension. 
For example itop-approval-portal or itop-standard-email-synchro

Before PR you'll get : 

image

After PR : 

image

I also changed the MissingDependencyException parent (from Exception to CoreException), as this is an iTop exception !

### CODE REVIEW

File: setup/modulediscovery.class.inc.php [hunk 1]
```diff
@@ -19,9 +19,42 @@
...
+class MissingDependencyException extends CoreException
 {
+	/**
+	 * @see \ModuleDiscovery::OrderModulesByDependencies property init
+	 * @var array module id as key
+	 *            another array as value, containing : 'module' with module info, 'dependencies' with missing dependencies
+	 */
 	public $aModulesInfo;
+
+	public function GetMessageForHtml()
```
Comment (Molkobain) [inline setup/modulediscovery.class.inc.php]: Some PHPDoc and `@since` could be cool

File: setup/modulediscovery.class.inc.php [hunk 2]
```diff
@@ -19,9 +19,42 @@
...
+	{
+		$sErrorMessage = <<<HTML
+<p>The following modules have unmet dependencies:</p>
+<ul>
+HTML;
+		foreach ($this->aModulesInfo as $sModuleId => $aModuleErrors) {
+			$sModuleLabel = $aModuleErrors['module']['label'];
+			$aModuleMissingDependencies = $aModuleErrors['dependencies'];
+			$sErrorMessage .= <<<HTML
+	<li><strong>{$sModuleLabel} ({$sModuleId})</strong>:
```
Comment (Molkobain) [inline setup/modulediscovery.class.inc.php]: I find it a bit hard to read having the impacted module name in bold. It seems to big to me and I would rather have it all in regular weight, the hierarchy is enough to understand, but not a blocking point to me.

**Comment by Molkobain** [comment]
Very nice initiative, should have been done so long ago... 😅😅😅

File: setup/modulediscovery.class.inc.php (hunk 2 above)
Comment (Hipska) [reply setup/modulediscovery.class.inc.php]: Maybe only the label in strong, and the module Id normal?

Comment (Molkobain) [reply setup/modulediscovery.class.inc.php]: Worth a try

Comment (piRGoif) [reply setup/modulediscovery.class.inc.php]: Without bold : 

image

Bold only on module label only: 

image

And module code only:
image

I like the second option ! 
Will commit that

Comment (piRGoif) [reply setup/modulediscovery.class.inc.php]: Done in 32896df9

Comment (Hipska) [reply setup/modulediscovery.class.inc.php]: I do like it as well, great!

File: setup/modulediscovery.class.inc.php (hunk 1 above)
Comment (piRGoif) [reply setup/modulediscovery.class.inc.php]: Done in a1cc8d1c

**Review by Hipska — APPROVED** [review]

**Comment by jbostoen** [comment]
Looks much nicer!

Is this also tackled in the CLI (unattended install)?

**Comment by piRGoif** [comment]
> Looks much nicer!

Nice to see people liking this change :)
 
> Is this also tackled in the CLI (unattended install)?

I'll have to take a look !
Actually I didn't change the original behavior : the exception message is still set as before, so other consumers calling $e->getMessage() will still see the same thing.

I'll look at the code and keep you posted !

File: setup/modulediscovery.class.inc.php (hunk 2 above)
Comment (Molkobain) [reply setup/modulediscovery.class.inc.php]: Great indeed 🙌

**Review by Molkobain — APPROVED** [review]

**Comment by piRGoif** [comment]
> Is this also tackled in the CLI (unattended install)?

The [unattended_install script](https://www.itophub.io/wiki/page?id=latest:advancedtopics:automatic_install) will call this stack :

```
\SetupUtils::CheckSelectedModules
   \ModuleDiscovery::GetAvailableModules
      \ModuleDiscovery::GetModules
         \ModuleDiscovery::OrderModulesByDependencies
```

The MissingDependencyException will be thrown in \ModuleDiscovery::OrderModulesByDependencies
It will be catched in \SetupUtils::CheckSelectedModules and the exception's message will be thrown in the returned $aResult array.

So, no change here.

File: setup/modulediscovery.class.inc.php [hunk 3]
```diff
@@ -19,9 +19,46 @@
...
+	 * @var array module id as key
+	 *            another array as value, containing : 'module' with module info, 'dependencies' with missing dependencies
+	 */
 	public $aModulesInfo;
+
+	/**
+	 * @return string HTML to print to the user the modules impacted
+	 * @since 2.7.7 3.0.2 3.1.0 PR #280
+	 */
+	public function GetMessageForHtml()
```
Comment (Molkobain) [inline setup/modulediscovery.class.inc.php]: Reviewed today, @piRGoif will rename the method to match `CoreException::getHtmlDesc()`

**Review by Molkobain — CHANGES_REQUESTED** [review]

File: setup/modulediscovery.class.inc.php (hunk 3 above)
Comment (piRGoif) [reply setup/modulediscovery.class.inc.php]: Done in b4840cca

**Review by Molkobain — APPROVED** [review]

**Comment by Molkobain** [comment]
Accepted in technical review.

**Comment by Molkobain** [comment]
Accepted in functional review, will be part of 2.7.7.
@piRGoif can you create the corresponding ticket and update the PR's title before we merge?

**Comment by piRGoif** [comment]
Done !
Merging...

**Pull request merged**